reinforced_lib/agents/deep/q_learning.py

```diff
@@ -0,0 +1,399 @@
+        """
+
+        x_dummy = jnp.empty(obs_space_shape)
+        params, state = q_network.init(key, x_dummy)
+
+        opt_state = optimizer.init(params)
+        replay_buffer = experience_replay.init()
+
+        return QLearningState(
+            params=params,
+            state=state,
+            opt_state=opt_state,
+            replay_buffer=replay_buffer,
+            prev_env_state=jnp.zeros(obs_space_shape),
+            epsilon=epsilon
+        )
+
+    @staticmethod
+    def loss_fn(
+            params: hk.Params,
+            key: PRNGKey,
+            state: hk.State,
+            params_target: hk.Params,
+            state_target: hk.State,
+            batch: Tuple,
+            q_network: hk.TransformedWithState,
+            discount: Scalar
+    ) -> Tuple[Scalar, hk.State]:
+        r"""
+        Loss is the Bellman error :math:`\mathcal{L}(\theta) = \mathbb{E}_{s, a, r, s'} \left[ \left( r + \gamma 
+        \max_{a'} Q(s', a') - Q(s, a) \right)^2 \right]` where :math:`s` is the current state, :math:`a` is the 
+        current action, :math:`r` is the reward, :math:`s'` is the next state, :math:`\gamma` is  the discount factor, 
+        :math:`Q(s, a)` is the Q-value of the state-action pair. Loss can be calculated on a batch of transitions.
+
+        Parameters
+        ----------
+        params : hk.Params
+            The parameters of the Q-network.
+        key : PRNGKey
+            A PRNG key used as the random key.
+        state : hk.State
+            The state of the Q-network.
+        params_target : hk.Params
+            The parameters of the target Q-network.
+        state_target : hk.State
+            The state of the target Q-network.
+        batch : Tuple
+            A batch of transitions from the experience replay buffer.
+        q_network : hk.TransformedWithState
+            The Q-network.
+        discount : Scalar
+            The discount factor.
+
+        Returns
+        -------
+        Tuple[Scalar, hk.State]
+            The loss and the new state of the Q-network.
+        """
+
+        states, actions, rewards, terminals, next_states = batch
+        q_key, q_target_key = jax.random.split(key)
+
+        q_values, state = q_network.apply(params, state, q_key, states)
+        q_values = jnp.take_along_axis(q_values, actions.astype(jnp.int32), axis=-1)
+
+        q_values_target, _ = q_network.apply(params_target, state_target, q_target_key, next_states)
+        target = rewards + (1 - terminals) * discount * jnp.max(q_values_target, axis=-1)
+
+        target = jax.lax.stop_gradient(target)
+        loss = jnp.square(target - jnp.squeeze(q_values)).mean()
+
+        return loss, state
+
+    @staticmethod
+    def update(
+            state: QLearningState,
+            key: PRNGKey,
+            env_state: Array,
+            action: Array,
+            reward: Scalar,
+            terminal: jnp.bool_,
+            q_network: hk.TransformedWithState,
```

Comment:
'q_network' is not used in this method. Is it on purpose as one of the arguments, or can it be deleted?

Reply:
Actually it is not needed, thanks for your attention!